File: zinc-flow-csharp/ZincFlow/Core/ContentStore.cs
using System.Buffers;
using System.Runtime.CompilerServices;

namespace ZincFlow.Core;

// --- ContentStore interface ---

public interface IContentStore
{
    string Store(byte[] data);
    byte[] Retrieve(string claimId);
    void Delete(string claimId);
    bool Exists(string claimId);
}

// --- FileContentStore: filesystem-backed with sharded directories ---

public sealed class FileContentStore : IContentStore
{
    private readonly string _baseDir;
    private long _claimCounter;

    public FileContentStore(string baseDir)
    {
        _baseDir = baseDir;
        Directory.CreateDirectory(baseDir);
    }

    public string Store(byte[] data)
    {
        var claimId = GenerateClaimId();
        var path = ClaimPath(claimId);
        File.WriteAllBytes(path, data);
        return claimId;
    }

    public byte[] Retrieve(string claimId)
    {
        var path = ClaimPath(claimId);
        if (!File.Exists(path)) return [];
        return File.ReadAllBytes(path);
    }

    public void Delete(string claimId)
    {
        var path = ClaimPath(claimId);
        if (File.Exists(path)) File.Delete(path);
    }

    public bool Exists(string claimId)
    {
        return File.Exists(ClaimPath(claimId));
    }

    private string ClaimPath(string claimId)
    {
        var prefix = claimId[..Math.Min(2, claimId.Length)];
        var dir = Path.Combine(_baseDir, prefix);
        Directory.CreateDirectory(dir);
        return Path.Combine(dir, claimId);
    }

    private string GenerateClaimId()
    {
        var counter = Interlocked.Increment(ref _claimCounter);
        return $"claim-{Environment.TickCount64}-{counter}";
    }
}

// --- MemoryContentStore: in-memory for testing ---

public sealed class MemoryContentStore : IContentStore
{
    private readonly Dictionary<string, byte[]> _data = new();
    private readonly object _lock = new();
    private long _counter;

    public string Store(byte[] data)
    {
        var id = $"mem-claim-{Interlocked.Increment(ref _counter)}";
        lock (_lock) { _data[id] = data; }
        return id;
    }

    public byte[] Retrieve(string claimId)
    {
        lock (_lock) { return _data.GetValueOrDefault(claimId) ?? []; }
    }

    public void Delete(string claimId)
    {
        lock (_lock) { _data.Remove(claimId); }
    }

    public bool Exists(string claimId)
    {
        lock (_lock) { return _data.ContainsKey(claimId); }
    }
}

// --- Content helpers ---

public static class ContentHelpers
{
    public const int ClaimThreshold = 256 * 1024;

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static Content MaybeOffload(IContentStore store, byte[] data)
    {
        if (data.Length > ClaimThreshold)
        {
            var claimId = store.Store(data);
            return new ClaimContent(claimId, data.Length);
        }
        return new Raw(data);
    }

    public static (byte[] Data, string Error) Resolve(IContentStore store, Content content)
    {
        if (content is Raw raw)
            return (raw.Data.ToArray(), "");
        if (content is ClaimContent claim)
            return (store.Retrieve(claim.ClaimId), "");
        if (content is RecordContent)
            return ([], "cannot resolve Records to raw bytes — use a RecordWriter");
        return ([], "unknown content type");
    }
}

File: zinc-flow-go/zinc-out/core/contentstore.go
package core

import (
	"os"
	"fmt"
	"time"
)

type ContentStore interface {
	Store(data []byte) string
	Retrieve(claimId string) []byte
	Delete(claimId string)
	Exists(claimId string) bool
}

//line /home/vrjoshi/proj/caravan-flow/zinc-flow-go/src/core/contentstore.zn:17
type FileContentStore struct {
	baseDir string
	claimCounter int
}

func NewFileContentStore(baseDir string) *FileContentStore {
	s := &FileContentStore{baseDir: baseDir, claimCounter: 0}
	err := os.MkdirAll(baseDir, 0o755)
	if err != nil {
		panic(err)
	}
	return s
}

func (s *FileContentStore) Store(data []byte) string {
	claimId := s.generateClaimId()
	path := s.claimPath(claimId)
	err1 := os.WriteFile(path, data, 0o644)
	if err1 != nil {
		panic(err1)
	}
	return claimId
}

func (s *FileContentStore) Retrieve(claimId string) []byte {
	path := s.claimPath(claimId)
	_tryerr2_val, _tryerr2_ret, _tryerr2 := func() ([]byte, bool, error) {
		data, err3 := os.ReadFile(path)
		if err3 != nil {
			return nil, false, err3
		}
		return data, true, nil
		return nil, false, nil
	}()
	if !_tryerr2_ret && _tryerr2 != nil {
		e := _tryerr2
		_ = e
		return make([]byte, 0)
	}
	if _tryerr2_ret {
		return _tryerr2_val
	}
	return nil
}

func (s *FileContentStore) Delete(claimId string) {
	path := s.claimPath(claimId)
	err4 := os.Remove(path)
	if err4 != nil {
		panic(err4)
	}
}

func (s *FileContentStore) Exists(claimId string) bool {
	path := s.claimPath(claimId)
	_, err := os.Stat(path)
	return err == nil
}

func (s *FileContentStore) claimPath(claimId string) string {
	prefix := claimId[0:2]
	dir := s.baseDir + "/" + prefix
	err5 := os.MkdirAll(dir, 0755)
	if err5 != nil {
		panic(err5)
	}
	return dir + "/" + claimId
}

func (s *FileContentStore) generateClaimId() string {
	s.claimCounter = s.claimCounter + 1
	return fmt.Sprintf("claim-%v-%v", time.Now().UnixNano(), s.claimCounter)
}


//line /home/vrjoshi/proj/caravan-flow/zinc-flow-go/src/core/contentstore.zn:69
type MemoryContentStore struct {
	data map[string][]byte
	counter int
}

func NewMemoryContentStore() *MemoryContentStore {
	return &MemoryContentStore{data: map[string][]byte{}, counter: 0}
}

func (s *MemoryContentStore) Store(content []byte) string {
	s.counter = s.counter + 1
	id := "mem-claim-" + fmt.Sprint(s.counter)
	s.data[id] = content
	return id
}

func (s *MemoryContentStore) Retrieve(claimId string) []byte {
	if func() bool { _, _ok := s.data[claimId]; return _ok }() {
		return s.data[claimId]
	}
	return make([]byte, 0)
}

func (s *MemoryContentStore) Delete(claimId string) {
	delete(s.data, claimId)
}

func (s *MemoryContentStore) Exists(claimId string) bool {
	return func() bool { _, _ok := s.data[claimId]; return _ok }()
}


//line /home/vrjoshi/proj/caravan-flow/zinc-flow-go/src/core/contentstore.zn:99
func MaybeOffload(store ContentStore, data []byte) Content {
	if len(data) > CONTENT_CLAIM_THRESHOLD {
		claimId := store.Store(data)
		return NewClaim(claimId, int64(len(data)))
	}
	return NewRaw(data)
}

//line /home/vrjoshi/proj/caravan-flow/zinc-flow-go/src/core/contentstore.zn:108
type ResolveResult struct {
	Data []byte
	ErrorMsg string
}

func NewResolveResult(data []byte, errorMsg string) ResolveResult {
	return ResolveResult{Data: data, ErrorMsg: errorMsg}
}

func (s ResolveResult) String() string {
	return fmt.Sprintf("ResolveResult(data=%v, errorMsg=%v)", s.Data, s.ErrorMsg)
}

//line /home/vrjoshi/proj/caravan-flow/zinc-flow-go/src/core/contentstore.zn:114
func Resolve(store ContentStore, c Content) ResolveResult {
	switch _v := c.(type) {
	case Raw:
		bytes := _v.Bytes
		return NewResolveResult(bytes, "")
	case Claim:
		claimId := _v.ClaimId
		return NewResolveResult(store.Retrieve(claimId), "")
	case Records:
		_ = _v
		return NewResolveResult(make([]byte, 0), "cannot resolve Records to raw bytes — use a RecordWriter")
	default:
		_ = _v
		panic("unreachable")
	}
	return ResolveResult{}
}

